data/finset.lean

```diff
@@ -623,6 +625,66 @@ finset.induction_on s ⟨0, empty_subset _⟩ $ λ a s ha ⟨n, hn⟩,
 
 end range
 
+/- Ico (a closed open interval) -/
+variables {n m l : ℕ}
+
+/-- `Ico n m` is the set of natural numbers `n ≤ k < m`. -/
+def Ico (n m : ℕ) : finset ℕ := ⟨_, nodup_Ico n m⟩
+
+namespace Ico
+
+@[simp] theorem val (n m : ℕ) : (Ico n m).1 = multiset.Ico n m := rfl
+
+@[simp] theorem mem : l ∈ Ico n m ↔ n ≤ l ∧ l < m := multiset.Ico.mem
+
+@[simp] theorem Ico_range (n : ℕ) : Ico 0 n = range n :=
+by dsimp [Ico]; congr; dsimp [multiset.Ico, list.Ico, multiset.range]; rw list.range_eq_range'
+
+@[simp] theorem empty (n : ℕ) : Ico n n = ∅ :=
+by dsimp [Ico]; congr; simp
+
+@[simp] theorem succ_singleton (n : ℕ) : Ico n (n+1) = singleton n :=
+by dsimp [Ico]; congr; simp
+
+lemma pred_singleton {m : ℕ} (h : m > 0) : Ico (m-1) m = {m-1} :=
+by dsimp [Ico]; congr; rw multiset.Ico.pred_singleton h; refl
+
+@[simp] lemma filter_lt (l n m : ℕ) : (Ico n m).filter (λ x, x < l) = (Ico n (min m l)) :=
```

Comment:
shouldn't this be proven for `list` and `multiset` first?

Reply:
Proving it for `list` feels really painful. I want to do it by first showing both lists are strictly ordered, and then showing that the result for finsets suffices!

Is there a sensible way to do it directly on lists, @johoelzl?

Reply:
Done now, per discussion below.